lib/pacemaker/pcmk_cluster_queries.c

```diff
@@ -0,0 +1,366 @@
+#include <glib.h>               // gboolean, GMainLoop, etc.
+#include <libxml/tree.h>        // xmlNode
+
+#include <pacemaker-internal.h>
+
+#include <crm/crm.h>
+#include <crm/cib.h>
+#include <crm/msg_xml.h>
+#include <crm/common/output_internal.h>
+#include <crm/common/xml.h>
+#include <crm/common/iso8601.h>
+#include <crm/common/ipc_controld.h>
+#include <crm/common/ipc_pacemakerd.h>
+#include <crm/common/mainloop.h>
+
+#include <crm/pacemaker/internal.h>
+
+#define DEFAULT_MESSAGE_TIMEOUT_MS 30000
+
+static guint message_timer_id = 0;
+
+static GMainLoop *mainloop = NULL;
+
+static crm_exit_t exit_code = CRM_EX_OK;
+
+static void
+start_main_loop(pcmk__output_t *out, guint message_timeout_ms)
+{
+    if (message_timeout_ms < 1) {
+        message_timeout_ms = DEFAULT_MESSAGE_TIMEOUT_MS;
+    }
+
+    exit_code = CRM_EX_DISCONNECT; // For unexpected disconnects
+    mainloop = g_main_loop_new(NULL, FALSE);
+    message_timer_id = g_timeout_add(message_timeout_ms,
+                                     admin_message_timeout,
+                                     GINT_TO_POINTER(message_timeout_ms));
+    g_main_loop_run(mainloop);
+}
+
+static void
+quit_main_loop(crm_exit_t ec)
+{
+    exit_code = ec;
+    if (mainloop != NULL) {
+        GMainLoop *mloop = mainloop;
+
+        mainloop = NULL; // Don't re-enter this block
+        pcmk_quit_main_loop(mloop, 10);
+        g_main_loop_unref(mloop);
+    }
+}
+
+gboolean
+admin_message_timeout(gpointer data)
+{
+    guint message_timeout_ms =  GPOINTER_TO_INT(data);
+
+    fprintf(stderr, "error: No reply received from controller before timeout (%dms)\n",
+            message_timeout_ms);
+    message_timer_id = 0;
+    quit_main_loop(CRM_EX_TIMEOUT);
+    return FALSE; // Tells glib to remove source
+}
+
+static void
+event_done(pcmk_ipc_api_t *api)
+{
+    pcmk_disconnect_ipc(api);
+    quit_main_loop(exit_code);
+}
+
+static pcmk_controld_api_reply_t *
+controld_event_reply(pcmk__output_t *out, pcmk_ipc_api_t *controld_api, enum pcmk_ipc_event event_type, crm_exit_t status, void *event_data)
+{
+    pcmk_controld_api_reply_t *reply = event_data;
+
+    switch (event_type) {
+        case pcmk_ipc_event_disconnect:
+            if (exit_code == CRM_EX_DISCONNECT) { // Unexpected
+                out->err(out, "error: Lost connection to controller");
+            }
+            event_done(controld_api);
+            return NULL;
+
+        case pcmk_ipc_event_reply:
+            break;
+
+        default:
+            return NULL;
+    }
+
+    if (message_timer_id != 0) {
+        g_source_remove(message_timer_id);
+        message_timer_id = 0;
+    }
+
+    if (status != CRM_EX_OK) {
+        out->err(out, "error: Bad reply from controller: %s",
+                crm_exit_str(status));
```

Comment:
FYI, IPC callbacks do get crm_exit_t (as an argument, not a return code) instead of standard return codes to indicate the API result, because API results potentially need to be valid on different hosts. crm_exit_t is defined entirely within Pacemaker, so its values are the same everywhere. Standard return codes can return errno, and errno can (and does) vary by OS etc. We do have places currently where we pass errno's between nodes, and those are landmines waiting for someone to hit them ... though thankfully, those are usually only used in log messages, so the worst case is someone gets a misleading log.